### DISCUSSION TITLE
Write BT into slave register.

### DISCUSSION THREAD

**Discussion by einarbjorkman** [artisan-roaster-scope/artisan#871] [Q&A] [answered]
Hello,
I am looking to create the following setup:
-I load/create a desired roast profile in Artisan
-Have an Arduino as a slave with the Modbus RTU library
-Write (with Artisan) the desired bean temperature into one of the Arduinos registers from
-Update the/send a new value at a specified frequency
Is this solution possible and, if so, how would I go about creating it?
All help appreciated!

-Einar

**Comment by MAKOMO** [comment, marked as answer]
Yes. You need to take advantage of the MODBUS external PID support of Artisan. Details are in the [blog post on the Artisan PID mechanism](https://artisan-roasterscope.blogspot.com/2016/11/pid-control.html).

Essentially, you need to define the SV register (the slave/register your client expects to receive the target temperature) in the MODBUS tabs PID section (menu `Config >> Ports`, 3rd tab). There you can also specify registers to set the p-i-d parameters of your client and commands to turn the PID loop on or off from within Artisan (eg. from its PID tab). In that PID tab you can then choose where this target temperature is taken from, either set manual/static (eg. via a SV slider or buttons) or taken from the background or a Ramp-Soak pattern.

**Reply by einarbjorkman** [reply]
Hi,
To be clear: I do not want to use the built in Artisan PID control system.
I will be using my own PID in the Arduino.
The only thing I want from Artisan is the SV / the desired BT.

"Essentially, you need to define the SV register (the slave/register your client expects to receive the target temperature) in the MODBUS tabs PID section (menu Config >> Ports, 3rd tab)."
So this would be refined in the [Registers, SV] box in the Ports >> Modbus tab?

"In that PID tab you can then choose where this target temperature is taken from, either set manual/static (eg. via a SV slider or buttons) or taken from the background or a Ramp-Soak pattern."
I do not see where to do this in the Port >> Modbus tab?

When you say "taken from the background", do you mean a predefined roast profile?

-Einar

**Reply by MAKOMO** [reply]
> Hi, To be clear: I do not want to use the built in Artisan PID control system. I will be using my own PID in the Arduino. The only thing I want from Artisan is the SV / the desired BT.

Understood

> 
> "Essentially, you need to define the SV register (the slave/register your client expects to receive the target temperature) in the MODBUS tabs PID section (menu Config >> Ports, 3rd tab)." So this would be refined in the [Registers, SV] box in the Ports >> Modbus tab?

As soon as there is a slave ID different from 0 set in the PID section of the MODBUS tab, the internal PID is disabled and the external PID is operated.

> 
> "In that PID tab you can then choose where this target temperature is taken from, either set manual/static (eg. via a SV slider or buttons) or taken from the background or a Ramp-Soak pattern." I do not see where to do this in the Port >> Modbus tab?

Oh, my bad. This is in the PID Dialog opened via the CONTROL button (which needs to be activated in the Device Dialog (menu Config >> Device, first tab, flag "Control").

> 
> When you say "taken from the background", do you mean a predefined roast profile?

Yes profile that got loaded into the background via the Background dialog (menu Config >> Background).

> 
> -Einar

**Reply by einarbjorkman** [reply]
Hi,
when you say "the external PID is operated", do you mean Artisan stops any of its own built in PID function?

![image](https://user-images.githubusercontent.com/104577724/166697679-89543bd9-a795-473e-a17d-ebbf5d7bd35b.png)

So that here making the slave=1 (in the PID section) effectively disables the Artisan PID??

and then...

![image](https://user-images.githubusercontent.com/104577724/166698079-f761c328-388b-4cb1-8c8a-9cce9cba3c53.png)

Here I just put the Set Value mode to Background? is that all?

I am also a bit confused why all of these options would be in a PID section if Artisan is not using PID?
Is it meant so that "this data will help an external PID" sort of?

-Einar

**Reply by MAKOMO** [reply]
> Hi, when you say "the external PID is operated", do you mean Artisan stops any of its own built in PID function?
> 

The duty is not calculated from the SV by the internal software PID if Artisan is configured to operate an external PID. This is then the responsibility of the external PID. It will receive from Artisan the SV and it can do whatever it like. For example it could compute itself a duty signal which it could send to the actor operating the burner.

> 
> So that here making the slave=1 (in the PID section) effectively disables the Artisan PID??
> 

Yes, and instructs Artisan to send the SV to that slave/register instead.

> and then...
> 
> 
> Here I just put the Set Value mode to Background? is that all?
> 

Yes

> I am also a bit confused why all of these options would be in a PID section if Artisan is not using PID? Is it meant so that "this data will help an external PID" sort of?

It is just one UI for both operating modes. Thus if you configure in the MODBUS tab also registers to set the parameters p-d-i, the settings in this PID dialog will be send to those registers in the external PID.

> 
> -Einar

**Reply by einarbjorkman** [reply]
Hmmm I guess I am unclear about the terminology:

"If Artisan is configured to operate an external PID" --> does this mean Artisan is meant to have som **control** over the Arduino's PID besides just providing the desired value/SV ?
(I realize you are saying below this that it's up to the Arduino but I want to be completely clear here...)

also what is the "duty" or "duty signal"? 

"It is just one UI for both operating modes. Thus if you configure in the MODBUS tab also registers to set the parameters p-d-i, the settings in this PID dialog will be send to those registers in the external PID."
OK, then I understand.

Thanks for continuous help- Einar

**Reply by MAKOMO** [reply]
> Hmmm I guess I am unclear about the terminology:
> 
> "If Artisan is configured to operate an external PID" --> does this mean Artisan is meant to have som **control** over the Arduino's PID besides just providing the desired value/SV ? (I realize you are saying below this that it's up to the Arduino but I want to be completely clear here...)

Yes, Artisan is just forwarding the SV value to the external PID.

> also what is the "duty" or "duty signal"?

What ever the PID computes. Did you read my post on the PID mechanism?

> 
> "It is just one UI for both operating modes. Thus if you configure in the MODBUS tab also registers to set the parameters p-d-i, the settings in this PID dialog will be send to those registers in the external PID." OK, then I understand.
> 
> Thanks for continuous help- Einar

**Reply by einarbjorkman** [reply]
Ok, so I have followed your instructions and attempted to provide a SV to the Arduino.
I import a CSV file of an example roast.
Use the "control" to configure the PID and then click "start" or "on".
However Artisan then prompts me to remove the preloaded profile (csv. file) or background, clears it, and then starts the recording.
How do I avoid this and just allow Artisan to follow the values of the uploaded profile??

-Einar

**Reply by MAKOMO** [reply]
You can only re-roast a background/template profile. Load the csv file, save as Artisan .alog profile. Load this one into the background.

**Reply by einarbjorkman** [reply]
hmmm, I am still not having success. Here are my exact settings and my procedure:
![image](https://user-images.githubusercontent.com/104577724/167872202-1532e844-b345-4bf7-948e-acd4741d0f02.png)
![image](https://user-images.githubusercontent.com/104577724/167872272-b9c44959-2037-447c-a7fd-5e707e5a185d.png)
![image](https://user-images.githubusercontent.com/104577724/167872343-46807053-53fe-4565-a40a-c06b10be3503.png)
![image](https://user-images.githubusercontent.com/104577724/167872434-3713402b-b378-4927-9117-2319d65aad85.png)
![image](https://user-images.githubusercontent.com/104577724/167872571-bacfa10f-ef7d-4ecc-b235-63ab871d2e6f.png)

I have tried initiating with both "start" and "begin", however the background profile is cleared in both cases and I'm reading the same value constantly in holding register [0].
Any suggestions?

-Einar

**Comment by einarbjorkman** [comment]
We have managed to solve it now, thank you for your help Marko!
It was done by "switch profiles", going to control and enabling "start PID on charge", and then clicking charge when started.
Besides the settings you helped with, that is :)
kind regards- Einar

**Reply by poundy** [reply]
thats because you loaded the profile, not loaded into background. Switch then moves it from the active profile to the background.

**Reply by poundy** [reply]
The two signs of that are the "profile background" dialog above does not have a filename at the bottom of it, and that the message on the last screenshot says the file was "loaded" not loaded into background.....

**Reply by einarbjorkman** [reply]
Ok, that makes sense.
Thank you for the info :)
Where would I select the option "load into background" directly without having to switch afterwards?

**Reply by poundy** [reply]
using the "load" button in the second last dialog box you showed, the "profile background" dialog. You can change background profile at any stage - although in your scenario that makes little sense since you are using that to follow the roast - and there's about 6 ways to get into that menu, Ctrl-B (for background) is my favourite way.

**Comment by fcebreros** [comment]
Hi,
it is possible.
https://lukasgrasse.medium.com/how-to-make-an-arduino-controlled-coffee-roaster-f6a3334fd7d5
there you can find info on how to do it.
Best
FC
